File: 006_ZigZagConversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
function convert(s, numRows) {
    let res = [];
    let step = 1, index = 0;
    for (let i = 0; i < s.length; i++) {
        if (res[index] === undefined) {
            res[index] = '';
        }
        res[index] += s[i];
        if (index === 0){
            step = 1;
        } else if (index === numRows - 1) {
            step = -1;
        }
        index += step;
    }
    return res.join('');
}

console.log(convert("PAYPALISHIRING", 3));
console.log(convert("PAYPALISHIRING", 4));
console.log(convert("AB", 1));

File: Java/006_ZigZagConversion.java
class Solution {
    public String convert(String s, int numRows) {
        char[] c = s.toCharArray();
        int len = c.length;
        StringBuffer[] sb = new StringBuffer[numRows];
        for (int i = 0; i < sb.length; i++) {
            sb[i] = new StringBuffer();
        }
        int i = 0;
        while (i < len) {
            for (int index = 0; index < numRows && i < len; index++) {
                sb[index].append(c[i++]);
            }
            for (int index = numRows - 2; i >= 1 && i < len; index--) {
                sb[index].append(c[i++]);
            }
        }
        for (int index = 1; index < sb.length; index++) {
            sb[0].append(sb[index]);
        }
        return sb[0].toString();
    }
}